ecs-agent/wsclient/client.go

```diff
@@ -0,0 +1,577 @@
+// The agent where the field is not applicable can be left empty
+type AgentConfig struct {
+	AWSRegion          string
+	AcceptInsecureCert bool
+	DockerEndpoint     string
+	IsDocker           bool
+}
+
+// ClientServerImpl wraps commonly used methods defined in ClientServer interface.
+type ClientServerImpl struct {
+	// cfg is the subset of user-specified runtime configuration
+	cfg *AgentConfig
+	// conn holds the underlying low-level websocket connection
+	conn wsconn.WebsocketConn
+	// CredentialProvider is used to retrieve AWS credentials
+	CredentialProvider *credentials.Credentials
+	// RequestHandlers is a map from message types to handler functions of the
+	// form:
+	//     "FooMessage": func(message *ecsacs.FooMessage)
+	RequestHandlers map[string]RequestHandler
+	// AnyRequestHandler is a request handler that, if set, is called on every
+	// message with said message. It will be called before a RequestHandler is
+	// called. It must take a single interface{} argument.
+	AnyRequestHandler RequestHandler
+	// MakeRequestHook is an optional callback that, if set, is called on every
+	// generated request with the raw request body.
+	MakeRequestHook MakeRequestHookFunc
+	// URL is the full url to the backend, including path, querystring, and so on.
+	URL string
+	// RWTimeout is the duration used for setting read and write deadlines
+	// for the websocket connection
+	RWTimeout time.Duration
+	// writeLock needed to ensure that only one routine is writing to the socket
+	writeLock sync.RWMutex
+	ClientServer
+	ServiceError
+	TypeDecoder
+}
+
+// MakeRequestHookFunc is a function that is invoked on every generated request
+// with the raw request body.  MakeRequestHookFunc must return either the body
+// to send or an error.
+type MakeRequestHookFunc func([]byte) ([]byte, error)
+
+// Connect opens a connection to the backend and upgrades it to a websocket. Calls to
+// 'MakeRequest' can be made after calling this, but responses will not be
+// receivable until 'Serve' is also called.
+func (cs *ClientServerImpl) Connect() error {
+	logger.Info("Establishing a Websocket connection", logger.Fields{
+		"url": cs.URL,
+	})
+	parsedURL, err := url.Parse(cs.URL)
+	if err != nil {
+		return err
+	}
+
+	wsScheme, err := websocketScheme(parsedURL.Scheme)
+	if err != nil {
+		return err
+	}
+	parsedURL.Scheme = wsScheme
+
+	// NewRequest never returns an error if the url parses and we just verified
+	// it did above
+	request, _ := http.NewRequest("GET", parsedURL.String(), nil)
+
+	// Sign the request; we'll send its headers via the websocket client which includes the signature
+	err = utils.SignHTTPRequest(request, cs.cfg.AWSRegion, ServiceName, cs.CredentialProvider, nil)
+	if err != nil {
+		return err
+	}
+
+	timeoutDialer := &net.Dialer{Timeout: wsConnectTimeout}
+	tlsConfig := &tls.Config{ServerName: parsedURL.Host, InsecureSkipVerify: cs.cfg.AcceptInsecureCert, MinVersion: tls.VersionTLS12}
+
+	//TODO: In order to get rid of the check -
```

Comment:
Yes, ideally I want to get rid of the check. I had a discussion with multiple folks and able to resolve few other such checks. I am preparing to doc to deal with these two checks. Once everyone agrees on the strategy, we can remove these checks as well. I had raised the PR so that ACS is unblocked, as driving consensus is taking some time.

Reply:
Sounds good, will look forward to this in a followup PR then.